File: 2020/11/Haskell/src/Main.hs
import           Data.Vector                    ( Vector )
import qualified Data.Vector                   as V
import           System.Environment
import           Text.Printf

type Grid = Vector (Vector Char)
type Point = (Int, Int)
type Bounds = (Int, Int)

gridInit :: [String] -> Grid
gridInit = V.fromList . map V.fromList

getSeat :: Point -> Grid -> Char
getSeat (r, c) = (V.! c) . (V.! r)

getBounds :: Grid -> Bounds
getBounds grid =
  (\[x, y] -> (x, y)) . map ((-1) +) $ [V.length grid, (V.length . V.head) grid]

updatePoint :: Point -> Char -> Grid -> Grid
updatePoint (r, c) x g = g'
 where
  row  = g V.! r
  row' = row V.// [(c, x)]
  g'   = g V.// [(r, row')]

getAdjacent :: Grid -> Point -> [Point]
getAdjacent g p@(px, py) = adjacent
 where
  (x, y) = getBounds g
  adjacent =
    [ (r, c)
    | r <- [px - 1 .. px + 1]
    , r >= 0
    , r <= x
    , c <- [py - 1 .. py + 1]
    , c >= 0
    , c <= y
    , (r, c) /= p
    ]

getInLineOfSight :: Grid -> Point -> [Point]
getInLineOfSight g (px, py) = lineOfSight
 where
  (x, y)    = getBounds g
  firstSeat = take 1 . dropWhile ((== '.') . (`getSeat` g))
  north     = firstSeat [ (r, py) | r <- [px - 1, px - 2 .. 0] ]
  south     = firstSeat [ (r, py) | r <- [px + 1 .. x] ]
  east      = firstSeat [ (px, c) | c <- [py + 1 .. y] ]
  west      = firstSeat [ (px, c) | c <- [py - 1, py - 2 .. 0] ]
  northWest = firstSeat
    [ (r, c)
    | r <- [px - 1, px - 2 .. 0]
    , c <- [py - 1, py - 2 .. 0]
    , (px - r) == (py - c)
    ]
  southEast = firstSeat
    [ (r, c) | r <- [px + 1 .. x], c <- [py + 1 .. y], (r - px) == (c - py) ]
  northEast = firstSeat
    [ (r, c)
    | r <- [px - 1, px - 2 .. 0]
    , c <- [py + 1 .. y]
    , (px - r) == (c - py)
    ]
  southWest = firstSeat
    [ (r, c)
    | r <- [px + 1 .. x]
    , c <- [py - 1, py - 2 .. 0]
    , (r - px) == (py - c)
    ]
  lineOfSight = concat
    [north, south, east, west, northWest, southEast, northEast, southWest]

stepMap :: Grid -> Grid -> Int -> [Point] -> Grid
stepMap _ g _ [] = g
stepMap g0 g n (p : ps)
  | this == '.'                        = stepMap g0 g n ps
  | this == 'L' && occupied == 0       = stepMap g0 (insert '#') n ps
  | this == '#' && occupied >= (n + 3) = stepMap g0 (insert 'L') n ps
  | otherwise                          = stepMap g0 g n ps
 where
  this = getSeat p g0
  seatFunction | n == 1    = getAdjacent
               | n == 2    = getInLineOfSight
               | otherwise = error $ "Unexpected value " ++ show n
  occupied = length . filter ((== '#') . (`getSeat` g0)) . seatFunction g $ p
  insert x = updatePoint p x g

cycleMap :: Grid -> Int -> [Point] -> Grid
cycleMap g n p | g' == g   = g
                 | otherwise = cycleMap g' n p
  where g' = stepMap g g n p

solve :: Int -> [String] -> Int
solve n x = length . V.concatMap (V.filter (== '#')) $ grid'
 where
  grid      = gridInit x
  (r, c)    = getBounds grid
  positions = [ (a, b) | a <- [0 .. r], b <- [0 .. c] ]
  grid'     = cycleMap grid n positions

output :: String -> IO ()
output path = do
  input <- filter (not . null) . lines <$> readFile path
  printf "File: %s\n" path
  printf "  Part 1: %d\n" . solve 1 $ input
  printf "  Part 2: %d\n" . solve 2 $ input

main :: IO ()
main = getArgs >>= mapM_ output

File: 2020/11/Rust/src/main.rs
use std::convert::TryFrom;
use std::env;
use std::fs;
use std::iter::repeat;

type Row = Vec<char>;
type Grid = Vec<Row>;
type Point = (usize, usize);

fn parse_input(input: &str) -> Grid {
    let input = input.lines().filter(|x| !x.is_empty()).collect::<Vec<_>>();
    input
        .iter()
        .map(|x| x.chars().collect::<Vec<_>>())
        .collect::<Vec<_>>()
}

fn get_seat(point: &Point, grid: &[Row]) -> char {
    let (x, y) = point;
    grid[*x][*y]
}

fn get_bounds(grid: &[Row]) -> Point {
    (grid.len() - 1, grid[0].len() - 1)
}

fn to_isize(num: &usize) -> isize {
    isize::try_from(*num).unwrap()
}

fn to_usize(num: &isize) -> usize {
    usize::try_from(*num).unwrap()
}

fn get_adjacent(grid: &[Row], point: &Point) -> Vec<Point> {
    let (x, y) = get_bounds(grid);
    let x = to_isize(&x);
    let y = to_isize(&y);
    let (px, py) = point;
    let px = to_isize(px);
    let py = to_isize(py);
    let mut adjacent = Vec::new();
    for r in (px - 1)..=(px + 1) {
        for c in (py - 1)..=(py + 1) {
            if (r >= 0)
                && (r <= x)
                && (c >= 0)
                && (c <= y)
                && ((r, c) != (px, py))
            {
                let qx = to_usize(&r);
                let qy = to_usize(&c);
                adjacent.push((qx, qy));
            }
        }
    }
    adjacent
}

fn first_seat<I>(line: I, grid: &[Row]) -> Option<Point>
where
    I: Iterator<Item = Point>,
{
    line.skip_while(|p| get_seat(p, grid) == '.').next()
}

fn get_in_line_of_sight(grid: &[Row], point: &Point) -> Vec<Point> {
    let (max_x, max_y) = get_bounds(grid);
    let (px, py) = point.to_owned();

    macro_rules! fs {
        ($i:expr) => {
            first_seat($i, grid)
        };
    }

    let n = fs!((0..px).rev().zip(repeat(py)));
    let s = fs!((px + 1..=max_x).zip(repeat(py)));
    let e = fs!(repeat(px).zip(py + 1..=max_y));
    let w = fs!(repeat(px).zip((0..py).rev()));

    let nw = fs!((0..px)
        .rev()
        .zip((0..py).rev())
        .filter(|(r, c)| px - r == py - c));

    let se = fs!((px + 1..=max_x)
        .zip(py + 1..=max_y)
        .filter(|(r, c)| r - px == c - py));

    let ne = fs!((0..px)
        .rev()
        .zip(py + 1..=max_y)
        .filter(|(r, c)| px - r == c - py));

    let sw = fs!((px + 1..=max_x)
        .zip((0..py).rev())
        .filter(|(r, c)| r - px == py - c));

    vec![n, s, e, w, nw, se, ne, sw]
        .iter()
        .filter(|p| p.is_some())
        .map(|p| p.unwrap())
        .collect::<Vec<_>>()
}

fn step_grid(mode: &usize, grid: &[Row], points: &[Point]) -> Grid {
    let mut result = grid.to_owned();
    for point in points {
        let this = get_seat(point, grid);
        let occupied = {
            if mode == &1 {
                get_adjacent(&grid, &point)
            } else {
                get_in_line_of_sight(&grid, &point)
            }
        };
        let occupied =
            occupied.iter().filter(|x| get_seat(x, grid) == '#').count();
        let (x, y) = point;
        if (this == 'L') && (occupied == 0) {
            result[*x][*y] = '#';
        } else if (this == '#') && (occupied >= mode + 3) {
            result[*x][*y] = 'L';
        }
    }
    result
}

fn cycle_grid(mode: &usize, grid: &[Row], points: &[Point]) -> Grid {
    let mut this = grid.to_owned();
    let mut next;
    loop {
        next = step_grid(mode, &this, points);
        if next == this {
            break;
        }
        this = next;
    }
    next
}

fn initial_positions(point: &Point) -> Vec<Point> {
    let mut positions = Vec::new();
    let (px, py) = *point;
    for x in 0..=px {
        for y in 0..=py {
            positions.push((x, y));
        }
    }
    positions
}

fn solve(mode: &usize, grid: &[Row]) -> usize {
    let bounds = get_bounds(grid);
    let positions = initial_positions(&bounds);
    let solved = cycle_grid(mode, grid, &positions);
    solved.concat().iter().filter(|x| *x == &'#').count()
}

fn output(path: &str) {
    let input = fs::read_to_string(&path).unwrap();
    let parsed = parse_input(&input);

    println!("File: {}", path);
    println!("  Part 1: {}", solve(&1, &parsed));
    println!("  Part 2: {}", solve(&2, &parsed));
}

fn main() {
    env::args().skip(1).for_each(|x| output(&x));
}
